src/controllers/scripting/controllerscriptmoduleengine.cpp

```diff
@@ -24,7 +24,7 @@ bool ControllerScriptModuleEngine::initialize() {
 
     QJSValue handleInputFunction = mod.property("handleInput");
     if (handleInputFunction.isCallable()) {
-        m_handleInputFunction = handleInputFunction;
+        m_handleInputFunction = wrapArrayBufferCallback(handleInputFunction);
```

Comment:
Not sure if this a great Idea to do unconditionally, IMO we should let the callback decide how it wants to handle the arraybuffer. And if they want to receive it as an `Uint8Array` every time, they can still decorate the handleInput function in the mapping itself.

Reply:
What would a use case be for a different view of an ArrayBuffer? If you really want this you can get the buffer with `Uint8Array.prototype.buffer`.

Reply:
well in the context of HID, it probably makes more sense to parse the buffer via a `DataView`, then when dealing with data from a Pot, it might make more sense to interpret the data as a signed integer than unsigned, etc. In both cases, we would be creating a Uint8Array just to throw it away shortly after.

Reply:
We need to do unconditionally in ControllerScriptEngineLegacy at least. How about for the new ControllerScriptModuleEngine we pass the ArrayBuffer to the fallback input callback and for the replacement for MidiDispatcher, those callbacks will get a Uint8Array?

Reply:
Now I remember our [prior discussion](https://mixxx.zulipchat.com/#narrow/stream/113295-controller-mapping/topic/HID.20(un-)packing) about handling HID data in JS. I agree for that case it would be better to pass the raw ArrayBuffer to the JS callback. Perhaps the ArrayBuffer could be more helpful for MIDI sysex messages too. But for regular 3 byte MIDI messages, I think those should be automatically converted to a Uint8Array.

Reply:
> How about for the new ControllerScriptModuleEngine we pass the ArrayBuffer to the fallback input callback and for the replacement for MidiDispatcher, those callbacks will get a Uint8Array?

Yes, though same here, IMO this should be in control of the mapper not any C++ hack. Its just as easy in JS than it is in C++ and its much cleaner imo:
```javascript
const decorateArrayBufConvert = func => (
    arraybuf => func(new Uint8Array(arraybuf))
);

export const handleInput = decorateArrayBufConvert((...) => {...})
```

Reply:
Why would you ever prefer an ArrayBuffer to a Uint8Array for a 3 byte MIDI message?

Reply:
> Why would you ever prefer an ArrayBuffer to a Uint8Array for a 3 byte MIDI message? 

Again, if you are dealing with a pot or encoder which uses signed messages, you don't have to manually convert the unsigned int into a signed one, but then can instead just use `DataView.getInt8(2)`

Why do we need to do the wrapping in the C++? I don't see any performance benefit and it just looks like a MidiController-specific hack. As I've shown earlier, the same can be accomplished purely in the mapping which results in a more similar API between HID/Bulk and Midi, and leaves more control to the user.

Reply:
signed ints in a MIDI message? I don't think I've ever heard of that. Is that even allowed by the MIDI standard?

Reply:
> it just looks like a MidiController-specific hack.

Right, that's why I proposed doing it just for MidiController but not others.

Reply:
> signed ints in a MIDI message? I don't think I've ever heard of that. Is that even allowed by the MIDI standard?

Depends. Raw midi always used unsigned 7-bit numbers. But Encoders may have a higher resolution that the 7 bit value range allows. In that case some manufacturers do something like this:

> The Data is a 14 bit signed or unsigned number split into 2 7-bit MIDI bytes, and sent LSB first.

Reply:
> signed ints in a MIDI message? I don't think I've ever heard of that. Is that even allowed by the MIDI standard?

Right, it isn't really. I forgot that its only possible to send 7-bit integers over the wire so it doesn't matter whether its accessed by `getUint8()` or `getInt8()`. You are right that the explicit casting doesn't make sense here. But you still haven't addressed my question why this has to be done in C++ and not in JS like I posted earlier.

Reply:
> Encoders may have a higher resolution that the 7 bit value range allows. In that case some manufacturers do something like this: The Data is a 14 bit signed or unsigned number split into 2 7-bit MIDI bytes, and sent LSB first.

Right, each message still needs to be converted to a Uint8Array in that case and the MSB and LSB need to be combined by the script with a little bitwise math.

> you still haven't addressed my question why this has to be done in C++ and not in JS like I posted earlier.

It doesn't *have* to be done in C++, but I don't understand why we wouldn't take care of this automatically for the user considering every 3 byte MIDI message needs to be converted to a Uint8Array per the MIDI standard. Otherwise the user has a bunch of boilerplate to deal with.

Reply:
IMO because it makes the API more opaque in the long run. I thought the idea was that we keep the basic wire API uniform on regardless on what technology is actually used and only add platform specific things such as the MidiDispatcher? And IMO 3-4 SLOC of boilerplate code are acceptable in this case. Automatically wrapping the arraybuffer in an `Uint8Array` just increases the amount of magic. Uniform API => less exceptions => less to remember for the mapper => better DX!

Reply:
> only add platform specific things such as the MidiDispatcher

Right, this would be a MIDI-specific thing only for MidiDispatcher.

Reply:
I'd be fine if we just do the conversion for the callbacks passed to `MidiDispatcher` as long as the default handler (`handleInput`) recieves the plain ArrayBuffer.

Reply:
That is exactly what I'm proposing.

Reply:
Oh, that wasn't apparent to me. The change I commented on was to wrap the modules `handleInput` function so I assumed we were talking about that. I'm sorry for wasting your time. As long as we only wrap the callbacks for MidiDispatcher but not the `handleInput` function I'm all for it.

Reply:
I changed this in f6f7050.

Reply:
> Right, this would be a MIDI-specific thing only for MidiDispatcher.

The dispatching part of the MidiDiapatchr will use the QByteArray directly, right? 

When a matching callback is found it can pass the Data as ArrayBuffer to JS. We may have registered the callback with a certain "decorateArrayBufConvert" converter for the typical use cases like Slider/button and so on in a way that the inner callback only needs to deal with the payload. 

That sounds neat.

Reply:
Yes, that's the idea @daschuer.